sklearn/feature_selection/tests/test_from_model.py

```diff
@@ -63,6 +64,83 @@ def test_input_estimator_unchanged():
     assert_true(transformer.estimator is est)
 
 
+def check_invalid_max_n_features(est, X, y):
+    max_n_features = X.shape[1]
+    for invalid_max_n_feature in [-1, max_n_features+1, 'gobbledigook']:
+        transformer = SelectFromModel(estimator=est,
+                                      max_n_features=invalid_max_n_feature)
+        assert_raises(ValueError, transformer.fit, X, y)
+
+
+def check_valid_max_n_features(est, X, y):
+    max_n_features = X.shape[1]
+    for valid_max_n_feature in [0, max_n_features, 'all', 5]:
+        transformer = SelectFromModel(estimator=est,
+                                      max_n_features=valid_max_n_feature)
+        transformer.fit(X, y)
+        X_new = transformer.transform(X)
+        if valid_max_n_feature == 'all':
+            valid_max_n_feature = max_n_features
+        assert_equal(X_new.shape[1], valid_max_n_feature)
+
+
+def check_max_n_features(est, X, y):
+    max_n_features = X.shape[1]
+
+    check_valid_max_n_features(est, X, y)
+    check_invalid_max_n_features(est, X, y)
+
+    transformer1 = SelectFromModel(estimator=est, max_n_features='all')
+    transformer2 = SelectFromModel(estimator=est,
+                                   max_n_features=max_n_features)
+    transformer1.fit(X, y)
+    transformer2.fit(X, y)
+    X_new1, X_new2 = transformer1.transform(X), transformer2.transform(X)
+    assert_array_equal(X_new1, X_new2)
+
+    """
+    Test max_n_features against actual model.
+    """
+    transformer1 = SelectFromModel(estimator=Lasso(alpha=0.025))
+    transformer1.fit(X, y)
+    X_new1 = transformer1.transform(X)
+    for n_features in range(1, X_new1.shape[1]+1):
+        transformer2 = SelectFromModel(estimator=Lasso(alpha=0.025),
+                                       max_n_features=n_features)
+        transformer2.fit(X, y)
+        X_new2 = transformer2.transform(X)
+        assert_array_equal(X_new1[:, :n_features], X_new2)
+
+    """
+    Test if max_n_features can break tie among feature importance
+    """
+    for n_features in range(max_n_features):
+        transformer = SelectFromModel(est, max_n_features=n_features)
+        X[:, n_features] = 1
```

Comment:
I'm not convinced. Firstly because you assume the estimator is perfect. A limited-depth tree will not select identical features the same number of times; it'll use one of them. But this still is not testing the behaviour under ties. You're much better contriving an example with fixed `feature_importances_`.